### ISSUE TITLE
no-extra-parens errors for ExpressionStatement restricted productions

### ISSUE THREAD

**Issue by michaelficarra** [eslint/eslint#4902]
Currently, `no-extra-parens` incorrectly reports

```js
(function(){});
```

and

```js
(class{});
```

and

```js
(let[a] = b);
```

as having unnecessary parens, even though they are needed. PR incoming.

**edit:** For more info, see the note here: http://tc39.github.io/ecma262/#prod-ExpressionStatement

**Comment by eslintbot** [comment]
@michaelficarra Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.
